File: internal/domain/interfaces.go
package domain

import (
	"context"
	"time"
)

// Metadata represents the audit record stored in the database.
type Metadata struct {
	S3Key     string    `json:"s3_key"`
	Timestamp time.Time `json:"timestamp"`
	Status    string    `json:"status"`
}

// APIClient defines the contract for fetching log entries from an external API.
type APIClient interface {
	FetchLogEntry(ctx context.Context) ([]byte, error)
}

// ObjectStorage defines the contract for storing raw log payloads.
type ObjectStorage interface {
	Upload(ctx context.Context, key string, data []byte) error
}

// Database defines the contract for persisting audit metadata.
type Database interface {
	SaveMetadata(ctx context.Context, metadata Metadata) error
}

File: vault_service/interfaces.py
from abc import ABC, abstractmethod
from typing import Dict, Any

class APIClient(ABC):
    @abstractmethod
    def fetch_log_entry(self) -> Dict[str, Any]:
        """Fetch log entry from external REST API."""
        pass

class ObjectStorage(ABC):
    @abstractmethod
    def upload_payload(self, key: str, payload: Dict[str, Any]) -> None:
        """Upload raw JSON payload to S3."""
        pass

class Database(ABC):
    @abstractmethod
    def write_metadata(self, metadata: Dict[str, Any]) -> None:
        """Write audit metadata into MongoDB."""
        pass
